### PR TITLE
Act-625: UI | Policy creation | Labels view | Filters

### PR DESCRIPTION
**Pull request by mmitoff**
Models necessary for ACT-625

### CODE REVIEW

File: src/sdk/models/metadataFilters.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+export enum MetadataConjunctionOperator {
+	And = 'and',
+	Or = 'or'
+}
+
+export enum MetadataPropertyFilterOperator {
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:6]: This doesn't look right and doesn't match the [server-side operators](https://github.com/portalarchitects/skysync-vnext/blob/198997fbe99590622e09930c2ad7572b52f288d7/Connectors/Connectors/src/PortalArchitects.Connectors.Primitives/Metadata/MetadataPropertyFilterOperator.cs#L11).

File: src/sdk/models/policies.ts [hunk 2]
```diff
@@ -73,7 +74,18 @@ export interface PolicyAuditEntry extends IEntityIdentifier<number> {
...
+}
+
 export interface PolicyTrackingGroup extends IEntityIdentifier<string>, IPrioritizedEntity {
 	name?: string;
+	description?: string;
+	priority?: number;
 	risk?: ContentRiskLevel;
+	rules?: any[];
+	assignment_rules?: PolicyTrackingGroupRule[];
+	entity_types?: any[];
```
Comment (smbecker) [inline src/sdk/models/policies.ts:89]:
```suggestion
	entity_types?: (EntityType & IPrioritizedEntity)[];
```

File: src/sdk/models/policies.ts [hunk 3]
```diff
@@ -73,7 +74,18 @@ export interface PolicyAuditEntry extends IEntityIdentifier<number> {
...
+	justification?: string;
+	criteria?: MetadataCalculatedFilter;
+}
+
 export interface PolicyTrackingGroup extends IEntityIdentifier<string>, IPrioritizedEntity {
 	name?: string;
+	description?: string;
+	priority?: number;
 	risk?: ContentRiskLevel;
+	rules?: any[];
```
Comment (smbecker) [inline src/sdk/models/policies.ts:87]:
```suggestion
```

File: src/sdk/models/policies.ts [hunk 4]
```diff
@@ -73,7 +74,18 @@ export interface PolicyAuditEntry extends IEntityIdentifier<number> {
...
 
+export interface PolicyTrackingGroupRule extends IEntityIdentifier<string> {
+	justification?: string;
+	criteria?: MetadataCalculatedFilter;
+}
+
 export interface PolicyTrackingGroup extends IEntityIdentifier<string>, IPrioritizedEntity {
 	name?: string;
+	description?: string;
+	priority?: number;
```
Comment (smbecker) [inline src/sdk/models/policies.ts:85]: Remove as this property comes through with `IPrioritizedEntity`.

**Review by smbecker — CHANGES_REQUESTED** [review]

File: src/sdk/models/metadataFilters.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+export interface MetadataPropertyFilter extends MetadataComponentFilter {
+	name?: string;
+	op?: MetadataPropertyFilterOperator;
+	value?: object;
+	select?: MetadataCalculatedExpression;
+}
+
+export interface MetadataScriptFilter extends MetadataComponentFilter {
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:60]: Remove as we don't actively support this at the moment

File: src/sdk/models/metadataFilters.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	In = 'In',
+	not_in = 'Not In',
+	NotIn = 'Not In',
+	empty = 'Empty',
+	Empty = 'Empty',
+	not_empty = 'Not Empty',
+	NotEmpty = 'Not Empty'
+}
+
+export interface MetadataComponentFilter { }
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:41]:
```suggestion
export type MetadataComponentFilter = MetadataConjunctionFilter | MetadataPropertyFilter;
```

Then remove the `extends MetadataComponentFilter` on the various filters

File: src/sdk/models/metadataFilters.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+export interface MetadataCalculatedFilter {
+	text: string;
+	expression: MetadataComponentFilter;
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:45]:
```suggestion
	text?: string;
	expression?: MetadataComponentFilter;
```

File: src/sdk/models/metadataFilters.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	op?: MetadataPropertyFilterOperator;
+	value?: object;
+	select?: MetadataCalculatedExpression;
+}
+
+export interface MetadataScriptFilter extends MetadataComponentFilter {
+	script: string;
+}
+
+export interface MetadataComponentExpression { }
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:64]:
```suggestion
export type MetadataComponentExpression = MetadataFunctionExpression | MetadataPropertyExpression MetadataStaticExpression;
```

File: src/sdk/models/metadataFilters.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+export interface MetadataFunctionExpression extends MetadataComponentExpression {
+	function: string;
+	args: MetadataComponentExpression[];
+}
+
+export interface MetadataPropertyExpression extends MetadataComponentExpression {
+	name: string;
+}
+
+export interface MetadataScriptExpression extends MetadataComponentExpression {
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:80]: Remove

File: src/sdk/models/metadataFilters.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+export interface MetadataComponentExpression { }
+
+export interface MetadataCalculatedExpression {
+	text: string;
+	expression: MetadataComponentExpression;
+}
+
+export interface MetadataFunctionExpression extends MetadataComponentExpression {
+	function: string;
+	args: MetadataComponentExpression[];
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:73]:
```suggestion
	args?: MetadataComponentExpression[];
```

File: src/sdk/models/metadataFilters.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+export interface MetadataScriptFilter extends MetadataComponentFilter {
+	script: string;
+}
+
+export interface MetadataComponentExpression { }
+
+export interface MetadataCalculatedExpression {
+	text: string;
+	expression: MetadataComponentExpression;
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:68]:
```suggestion
	text?: string;
	expression?: MetadataComponentExpression;
```

File: src/sdk/models/metadataFilters.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+export interface MetadataPropertyExpression extends MetadataComponentExpression {
+	name: string;
+}
+
+export interface MetadataScriptExpression extends MetadataComponentExpression {
+	script: string;
+}
+
+export interface MetadataStaticExpression extends MetadataComponentExpression {
+	value: string;
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:85]:
```suggestion
	value: any;
```

**Review by smbecker — CHANGES_REQUESTED** [review]

File: src/sdk/models/metadataFilters.ts (hunk 1 above)
Comment (mmitoff) [reply src/sdk/models/metadataFilters.ts:6]: I began with just the server side enum values for interpreting the "op" values on the metadata property filters.  I ended up adding the enum members as values as well to handle interpreting the "function" values on metadata function expressions. I could move the enum member values "Equals, NotEqual, etc" into a different enum?

Comment (smbecker) [reply src/sdk/models/metadataFilters.ts:6]: This enum should just be the server side enum values. Anything else should be a separate enum. Also, I don't recall us having function names like 'Equals' and 'NotEquals'. The possible function names comes from the registered function provides, which is defaulted [here](https://github.com/portalarchitects/skysync-vnext/blob/develop/Connectors/Connectors/src/PortalArchitects.Connectors.Transfers/Metadata/MetadataExpressionParser.cs#L28). You could make the default functions into an enum and then change `MetadataFunctionExpression.name` to `MetadataFunctionName | string`.

Comment (mmitoff) [reply src/sdk/models/metadataFilters.ts:6]: Refactored

File: src/sdk/models/metadataFilters.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
+export enum MetadataConjunctionOperator {
+	And = 'and',
+	Or = 'or'
+}
+
+export enum MetadataPropertyFilterOperator {
+	eq = 'Equals',
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:7]: These should actually be reversed (i.e. `Equals = 'eq'`).

File: src/sdk/models/metadataFilters.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	empty = 'Empty',
+	not_empty = 'Not Empty',
+}
+
+export enum MetadataFunctionName {
+	Now = 'Now',
+	Year = 'Year',
+	Month = 'Month',
+	Day = 'Day',
+	AddDays = 'Add Days',
```
Comment (smbecker) [inline src/sdk/models/metadataFilters.ts:30]: The part in quotes should match the function name, which shouldn't have any spaces in it.

**Review by smbecker — CHANGES_REQUESTED** [review]

File: src/sdk/models/metadataFilters.ts (hunk 13 above)
Comment (mmitoff) [reply src/sdk/models/metadataFilters.ts:7]: Should I update the enum member name as well then (change 'Less Than' or LessThan) or leave as is:

export enum MetadataPropertyFilterOperator {
	Equals = 'eq',
	'Not Equal'= 'ne',
	'Less Than' = 'lt',
	'Less Than Equal To' = 'le',
	'Greater Than' = 'gt',
	'Greater Than Equal To' = 'ge',
	'Starts With' = 'starts_with',
	'Does Not Start With' = 'not_starts_with',
	'Ends With' = 'ends_with',
	'Does Not End With' = 'not_ends_with',
	Contains = 'contains',
	'Does Not Contain' = 'not_contains',
	In = 'in',
	'Not In' = 'not_in',
	Empty = 'empty',
	'Not Empty' = 'not_empty',
}

Comment (mmitoff) [reply src/sdk/models/metadataFilters.ts]: Updated

File: src/sdk/models/metadataFilters.ts (hunk 14 above)
Comment (mmitoff) [reply src/sdk/models/metadataFilters.ts]: Updated

**Review by smbecker — APPROVED** [review]

**Pull request merged**